### PR TITLE
[SPARK-18702][SQL] input_file_block_start and input_file_block_length

### PR DESCRIPTION
**Pull request by rxin**
## What changes were proposed in this pull request?
We currently have function input_file_name to get the path of the input file, but don't have functions to get the block start offset and length. This patch introduces two functions:

1. input_file_block_start: returns the file block start offset, or -1 if not available.

2. input_file_block_length: returns the file block length, or -1 if not available.

## How was this patch tested?
Updated existing test cases in ColumnExpressionSuite that covered input_file_name to also cover the two new functions.

### CODE REVIEW

**Comment by rxin** [comment]
cc @ueshin

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 1]
```diff
@@ -533,31 +533,54 @@ class ColumnExpressionSuite extends QueryTest with SharedSQLContext {
...
       val data = sparkContext.parallelize(0 to 10).toDF("id")
       data.write.parquet(dir.getCanonicalPath)
-      val answer = spark.read.parquet(dir.getCanonicalPath).select(input_file_name())
-        .head.getString(0)
-      assert(answer.contains(dir.getCanonicalPath))
 
-      checkAnswer(data.select(input_file_name()).limit(1), Row(""))
+      // Test the 3 expressions when reading from files
+      val q = spark.read.parquet(dir.getCanonicalPath).select(
+        input_file_name(), expr("input_file_block_start()"), expr("input_file_block_length()"))
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: We should add `input_file_block_start()`, `input_file_block_length()` to `functions.scala` and use them the same as `input_file_name()`?

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 2]
```diff
@@ -533,31 +533,54 @@ class ColumnExpressionSuite extends QueryTest with SharedSQLContext {
...
       val data = sparkContext.parallelize((0 to 10).map(_.toString)).toDF()
       data.write.text(dir.getCanonicalPath)
       val df = spark.sparkContext.textFile(dir.getCanonicalPath).toDF()
-      val answer = df.select(input_file_name()).head.getString(0)
-      assert(answer.contains(dir.getCanonicalPath))
 
-      checkAnswer(data.select(input_file_name()).limit(1), Row(""))
+      // Test the 3 expressions when reading from files
+      val q = df.select(
+        input_file_name(), expr("input_file_block_start()"), expr("input_file_block_length()"))
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: ditto.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala [hunk 3]
```diff
@@ -567,10 +590,22 @@ class ColumnExpressionSuite extends QueryTest with SharedSQLContext {
         classOf[LongWritable],
         classOf[Text])
       val df = rdd.map(pair => pair._2.toString).toDF()
-      val answer = df.select(input_file_name()).head.getString(0)
-      assert(answer.contains(dir.getCanonicalPath))
 
-      checkAnswer(data.select(input_file_name()).limit(1), Row(""))
+      // Test the 3 expressions when reading from files
+      val q = df.select(
+        input_file_name(), expr("input_file_block_start()"), expr("input_file_block_length()"))
```
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: ditto.

File: core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala [hunk 4]
```diff
@@ -132,54 +132,57 @@ class NewHadoopRDD[K, V](
...
       logInfo("Input split: " + split.serializableHadoopSplit)
-      val conf = getConf
+      private val conf = getConf
 
-      val inputMetrics = context.taskMetrics().inputMetrics
-      val existingBytesRead = inputMetrics.bytesRead
+      private val inputMetrics = context.taskMetrics().inputMetrics
+      private val existingBytesRead = inputMetrics.bytesRead
 
       // Sets the thread local variable for the file's name
```
Comment (ueshin) [inline core/src/main/scala/org/apache/spark/rdd/NewHadoopRDD.scala]: I agree with it and also a comment in `HadoopRDD`.

File: sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala (hunk 1 above)
Comment (ueshin) [inline sql/core/src/test/scala/org/apache/spark/sql/ColumnExpressionSuite.scala]: I see.

File: core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+   * @param length size of the block, in bytes, or -1 if not available.
+   */
+  private class FileBlock(val filePath: UTF8String, val startOffset: Long, val length: Long) {
+    def this() {
+      this(UTF8String.fromString(""), -1, -1)
+    }
+  }
+
+  /**
+   * The thread variable for the name of the current file being read. This is used by
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/rdd/InputFileBlockHolder.scala]: nit: this comment can be updated too.

**Comment by rxin** [comment]
Thanks - going to merge this in master.

**Pull request closed without merging**